### ISSUE TITLE
Mulitask model with to extra heads and RSConv as common backbone

### ISSUE THREAD

**Issue by ttsesm** [nicolas-chaulet/torch-points3d#620]
Hi @nicolas-chaulet,

I would like to ask whether it is possible to have a multitask model (classification+regression) where I will create two extra heads which will address each task independently but having the RSConv model as the common backbone. At the moment for addressing such a task I was having a multitask loss (combined losses), through the following model:

```
class RSConvCLassifier(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.encoder = RSConv("unet", input_nc=8, output_nc=3, num_layers=4)

    @property
    def conv_type(self):
        """ This is needed by the dataset to infer which batch collate should be used"""
        return self.encoder.conv_type

    def forward(self, data):
        return self.encoder(data)
```

which has 3 output channels, 1 for regression and 2 for classification (it is a 2-classes binary classification problem) then in my loss function I was combining the two outputs:

```
class MultiTaskLoss(nn.Module):
    def __init__(self):
        super(MultiTaskLoss, self).__init__()
        task_num = 2
        self.task_num = task_num
        self.log_vars = nn.Parameter(torch.zeros((task_num)))

    def forward(self, gt, preds):
        l1, ce = nn.L1Loss(), nn.CrossEntropyLoss()

        y = einops.rearrange(preds, 'd0 d1 d2 -> (d0 d2) d1')
        x = einops.rearrange(gt, 'd0 d1 d2 -> (d0 d1) d2')

        loss0 = l1(y[:, 0], x[:, 0])
        loss1 = ce(y[:,1:3], x[:,1].to(int))

        precision0 = torch.exp(-self.log_vars[0])
        loss0 = precision0 * loss0 + self.log_vars[0]

        precision1 = torch.exp(-self.log_vars[1])
        loss1 = precision1 * loss1 + self.log_vars[1]

        return loss0+loss1
```

Regarding how to create the to extra heads, my guess would be something like:

```
class RSConvCLassifier(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.encoder = RSConv("unet", input_nc=8, output_nc=3, num_layers=4)
        self.head1 = torch.nn.Linear(self.encoder.output_nc[0], 1) # output layer for regression
        self.head2 = torch.nn.Linear(self.encoder.output_nc[1:2], 2) # output layer for classification

    @property
    def conv_type(self):
        """ This is needed by the dataset to infer which batch collate should be used"""
        return self.encoder.conv_type

    def forward(self, data):
        return self.encoder(data)
```
but I am not sure whether this is correct. Then I am not sure how to address the loss for each head, should I have an individual loss for each head which I back propagate? Thus, if you have any experience I would be happy to hear your feedback.

Thanks.

**Comment by humanpose1** [comment]
I'm not an expert of Multi task learning. However, I think you should have a loss for each  and you apply the loss for each head and then you backpropagate the total loss. The problem with your current implementation is that you cannot track each loss individually.

**Comment by ttsesm** [comment]
> I'm not an expert of Multi task learning. However, I think you should have a loss for each and you apply the loss for each head and then you backpropagate the total loss. The problem with your current implementation is that you cannot track each loss individually.

Well your comment makes sense, however there are two points that I am trying to figure out: 
1) What is the correct way to create the two additional heads and 
2) also related to your comment how apply this different losses (I am not sure whether it needs to be combined to a total loss though) corresponding to each head accordingly. In principle I could have something like:

```
logit1 = network(X1) # first stream of data
loss_1 = compute_loss_1(logit1)
loss_1.backward()
optimizer_1.step()

logit2 = network(X2) # second stream of data
loss_2 = compute_loss_2(logit2)
loss_2.backward()
optimizer_2.step()
```

but again I am not sure and this is what I am trying to understand.

**Comment by nicolas-chaulet** [comment]
If I understand correctly, you have a common backbone and multiple heads that you want to supervise with different losses. The way I've done it (for object detection for example) is to compute all the losses and add them together at the end with some weighting to make sure they have a similar amplitude. The code you posted initially looks good in principle, the gradient should be computed on the aggregate loss. You can take a look at Votenet for example, we use like 5 or 6 losses in that example.

**Comment by ttsesm** [comment]
Thanks @nicolas-chaulet, is it easy to point me to the exact link of the Votenet example that you are reffering to so that I can have a look because looking into the examples folder I do not see anything related to Votenet.

**Comment by CCInc** [comment]
Take a look at: https://github.com/nicolas-chaulet/torch-points3d/blob/master/torch_points3d/models/object_detection/votenet2.py

Or, you can look at this file, where I use multiple losses: https://github.com/nicolas-chaulet/torch-points3d/blob/faa0c88024891fd80b9a6ce750e58d972e8f8901/torch_points3d/models/classification/point_augment.py

**Comment by ttsesm** [comment]
@CCInc following the links that you've posted I've tried to create the following model:

```
class RSConvCLassifier(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.backbone_model = RSConv("unet", input_nc=8, output_nc=3, num_layers=4)

        # REGRESSION HEAD
        self.head1 = (
            Seq()
                .append(MLP([self.backbone_model.output_nc, self.backbone_model.output_nc], bias=False))
                .append(torch.nn.Linear(self.backbone_model.output_nc, 1))
        )

        # CLASSIFICATION HEAD
        self.head2 = (
            Seq()
                .append(MLP([self.backbone_model.output_nc, self.backbone_model.output_nc], bias=False))
                .append(torch.nn.Linear(self.backbone_model.output_nc, 2))
                # .append(torch.nn.LogSoftmax())
        )

    @property
    def conv_type(self):
        """ This is needed by the dataset to infer which batch collate should be used"""
        return self.backbone_model.conv_type

    def forward(self, data):
        # return self.backbone_model(data)
        return self.head1(data), self.head2(data)
```

However, this gives me the following error: `{TypeError}linear(): argument 'input' (position 1) must be Tensor, not SimpleBatch` any idea what is wrong?

**Comment by humanpose1** [comment]
you should do `self.head1(data.x)`

**Comment by ttsesm** [comment]
> you should do `self.head1(data.x)`

hhmm this gives me `{RuntimeError}mat1 dim 1 must match mat2 dim 0 mat1 dim1:8 mat2 dim0: 3` if I check on the output of the `self.backbone(data)` that I was using before the output is: `Data(pos=[10, 8192, 3], x=[10, 3, 8192])`

![image](https://user-images.githubusercontent.com/10189018/122745042-e25c0a00-d288-11eb-87e3-6576edde49b2.png)

**Comment by ttsesm** [comment]
ok, it seems that my `MLP([self.backbone_model.output_nc, self.backbone_model.output_nc], bias=False)` needs to be changed to  `MLP([data.x.shape[2], self.backbone_model.output_nc], bias=False)` where `data.x.shape[2] = 8`

**Comment by ttsesm** [comment]
Do you see anything wrong in the following training for loop, because my output is crappy:
```
class RSConvCLassifier(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.backbone_model = RSConv("unet", input_nc=8, output_nc=3, num_layers=4)

        # REGRESSION HEAD
        self.head1 = (
            Seq()
                .append(MLP([8, self.backbone_model.output_nc], bias=False))
                .append(torch.nn.Linear(self.backbone_model.output_nc, 1))
        )

        # CLASSIFICATION HEAD
        self.head2 = (
            Seq()
                .append(MLP([8, self.backbone_model.output_nc], bias=False))
                .append(torch.nn.Linear(self.backbone_model.output_nc, 2))
                # .append(torch.nn.LogSoftmax())
        )

    @property
    def conv_type(self):
        """ This is needed by the dataset to infer which batch collate should be used"""
        return self.backbone_model.conv_type

    def forward(self, data):
        # return self.backbone_model(data)
        return self.head1(data.x), self.head2(data.x)
```

```
def train_epoch(model, loss_fn1, loss_fn2, optimizer, data_loader, device, results_dir=None):
    model.train()
    accumulation_steps = 1
    losses1 = []
    losses2 = []

    with Ctq(data_loader) as tq_data_loader:
        for i, data in enumerate(tq_data_loader):
            data.to(device)

            head1_data_out, head2_data_out = model.forward(data)
            output1 = head1_data_out.squeeze()
            output2 = head2_data_out.squeeze()
            labels1 = data.y[:,:,0].squeeze()
            labels2 = data.y[:,:,1].squeeze()

            loss1 = loss_fn1(labels1, output1) / accumulation_steps # loss_fn1 = nn.L1Loss()
            loss2 = loss_fn2(labels2.to(int), output2) / accumulation_steps # and loss_fn2 = nn.CrossEntropyLoss()

            losses1.append(float(loss1))
            losses2.append(float(loss2))

            loss1.backward()
            loss2.backward()

            if (i - 1) % accumulation_steps == 0:
                optimizer.step()
                # Reset gradients, for the next accumulated batches
                optimizer.zero_grad()

```
where my `model = RSConvCLassifier()`. I am not sure whether I can back propagate the two losses the way I am doing it at the moment because I saw [here](https://github.com/nicolas-chaulet/torch-points3d/blob/faa0c88024891fd80b9a6ce750e58d972e8f8901/torch_points3d/models/classification/point_augment.py#L62) that `freeze()/unfreeze()` is used respectively but there are also two models instead of one that I have in my case.

**Comment by ttsesm** [comment]
Just for the record to mention that modifying my model as follows did the trick:

```
class RSConvCLassifier(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.backbone_model = RSConv("unet", input_nc=8, output_nc=3, num_layers=4)

        # self.head1 = torch.nn.Linear(self.backbone_model.output_nc, 1)
        # self.head2 = torch.nn.Linear(self.backbone_model.output_nc, 2)

        # REGRESSION HEAD
        self.head1 = (
            Seq()
                .append(MLP([self.backbone_model.output_nc, self.backbone_model.output_nc], bias=False))
                .append(torch.nn.Linear(self.backbone_model.output_nc, 1))
        )

        # CLASSIFICATION HEAD
        self.head2 = (
            Seq()
                .append(MLP([self.backbone_model.output_nc, self.backbone_model.output_nc], bias=False))
                .append(torch.nn.Linear(self.backbone_model.output_nc, 2))
        )

    @property
    def conv_type(self):
        """ This is needed by the dataset to infer which batch collate should be used"""
        return self.backbone_model.conv_type


    def forward(self, data):
        data_features = self.backbone_model(data)

        backbone_feats = data_features.x.clone()
        self.semantic_labels = data.y[:,:,0].flatten()

        if backbone_feats.dim() == 3:
            backbone_feats = backbone_feats.transpose(2, 1)
            backbone_feats = backbone_feats.reshape(-1, backbone_feats.shape[2])
        if backbone_feats.shape[0] == self.semantic_labels.shape[0]:
            return self.head1(backbone_feats), self.head2(backbone_feats)
```

and then of course during the training I had to use the `retain_graph=True` flag.

```
loss1.backward(retain_graph=True)
loss2.backward()
```
